File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.Stack;
class Solution {
    public int solution(String s) {
        int answer = 0;
        int N = s.length();
        for (int i = 0; i < N; i++) {
            String tmp = s.substring(i, N) + s.substring(0, i);
            boolean possible = true;
            Stack<Character> stack = new Stack<>();
            for (int j = 0; j < N; j++) {
                if (tmp.charAt(j) == '{' || tmp.charAt(j) == '(' || tmp.charAt(j) == '[') {
                    stack.push(tmp.charAt(j));
                } else if (tmp.charAt(j) == '}') {
                    if (!stack.isEmpty() && stack.peek() == '{') {
                        stack.pop();
                    } else {
                        possible = false;
                        break;
                    }
                } else if (tmp.charAt(j) == ')') {
                    if (!stack.isEmpty() && stack.peek() == '(') {
                        stack.pop();
                    } else {
                        possible = false;
                        break;
                    }
                } else if (tmp.charAt(j) == ']') {
                    if (!stack.isEmpty() && stack.peek() == '[') {
                        stack.pop();
                    } else {
                        possible = false;
                        break;
                    }
                }
            }
            
            if (possible && stack.isEmpty()) {
                answer++;
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    answer = 0
    N = len(s)
    for i in range(N):
        tmp = s[i:] + s[:i]
        stack = []
        possible = True
        for i in range(N):
            if tmp[i] == '{' or tmp[i] == '(' or tmp[i] == '[':
                stack.append(tmp[i])
            elif tmp[i] == '}':
                if stack and stack[-1] == '{':
                    stack.pop()
                else:
                    possible = False
                    break
            elif tmp[i] == ')':
                if stack and stack[-1] == '(':
                    stack.pop()
                else:
                    possible = False
                    break
            elif tmp[i] == ']':
                if stack and stack[-1] == '[':
                    stack.pop()
                else:
                    possible = False
                    break
                    
        if possible and not stack:
            answer += 1
    return answer